src/core/radio/ble_secure.cpp

```diff
@@ -0,0 +1,540 @@
+
+    SuccessOrAssert(mTls.SetPsk(reinterpret_cast<const uint8_t *>(aPskd.GetAsCString()), aPskd.GetLength()));
+}
+
+#if defined(MBEDTLS_BASE64_C) && defined(MBEDTLS_SSL_KEEP_PEER_CERTIFICATE)
+Error BleSecure::GetPeerCertificateBase64(unsigned char *aPeerCert, size_t *aCertLength)
+{
+    Error error;
+
+    VerifyOrExit(aCertLength != nullptr, error = kErrorInvalidArgs);
+
+    error = mTls.GetPeerCertificateBase64(aPeerCert, aCertLength, *aCertLength);
+
+exit:
+    return error;
+}
+#endif // defined(MBEDTLS_BASE64_C) && defined(MBEDTLS_SSL_KEEP_PEER_CERTIFICATE)
+
+Error BleSecure::SendMessage(ot::Message &aMessage)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(IsConnected(), error = kErrorInvalidState);
+    VerifyOrExit(mSentMessage != nullptr, error = kErrorNoBufs);
+    SuccessOrExit(error = mSentMessage->AppendBytesFromMessage(aMessage, 0, aMessage.GetLength()));
+
+    if (mSentMessage->GetLength() > kFlushThreshold)
+    {
+        error = Flush();
+    }
+
+    if (error == kErrorNone)
+    {
+        aMessage.Free();
+    }
+
+exit:
+    return error;
+}
+
+Error BleSecure::Send(uint8_t *aBuf, uint16_t aLength)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(IsConnected(), error = kErrorInvalidState);
+    VerifyOrExit(mSentMessage != nullptr, error = kErrorNoBufs);
+    SuccessOrExit(error = mSentMessage->AppendBytes(aBuf, aLength));
+
+    if (mSentMessage->GetLength() > kFlushThreshold)
+    {
+        error = Flush();
+    }
+
+exit:
+    return error;
+}
+
+Error BleSecure::SendApplicationTlv(uint8_t *aBuf, uint16_t aLength)
+{
+    if (aLength > 254)
+    {
+        ot::ExtendedTlv tlv;
+
+        tlv.SetType(ot::MeshCoP::TcatAgent::kTlvSendApplicationData);
+        tlv.SetLength(aLength);
+        Send(reinterpret_cast<uint8_t *>(&tlv), sizeof(tlv));
+    }
+    else
+    {
+        ot::Tlv tlv;
+
+        tlv.SetType(ot::MeshCoP::TcatAgent::kTlvSendApplicationData);
+        tlv.SetLength(aLength);
+        Send(reinterpret_cast<uint8_t *>(&tlv), sizeof(tlv));
+    }
+
+    return Send(aBuf, aLength);
+}
+
+Error BleSecure::Flush(void)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(IsConnected(), error = kErrorInvalidState);
+    VerifyOrExit(mSentMessage != nullptr, error = kErrorNoBufs);
+    VerifyOrExit(mSentMessage->GetLength() != 0, error = kErrorNone);
+
+    mTransmitQueue.Enqueue(*mSentMessage);
+    mTransmitTask.Post();
+
+    mSentMessage = Get<MessagePool>().Allocate(Message::kTypeBle);
+    VerifyOrExit(mSentMessage != nullptr, error = kErrorNoBufs);
+
+exit:
+    return error;
+}
+
+Error BleSecure::HandleBleReceive(uint8_t *aBuf, uint16_t aLength)
+{
+    ot::Message     *message = nullptr;
+    Ip6::MessageInfo theMessageInfo;
+
+    VerifyOrExit((message = Get<MessagePool>().Allocate(Message::kTypeBle, 0)) != nullptr);
+    SuccessOrExit(message->AppendBytes(aBuf, aLength));
+
+    // Cannot call Receive(..) directly because Setup(..) and mState are private
```

Comment:
I think that handling the state machine in two places isn't worth the effort. Let's leave it as it is.

Reply:
Just wondering if the name `HandleUdpReceive` is still correct? Now it's generalized to SecureTransport it can presumably also accept TLS-over-TCP, and TLS-over-BLE(TCAT), so not just UDP for DTLS-over-UDP.

Maybe `HandleMsgReceive` ?

Reply:
@abtink Maybe you have an idea on this: since the DTLS is now made generic into "secure transport", it seems not correct to have the handler named `HandleUdpReceive` . Since it could be also via TCP. Or TLS via BLE, as in our case, which doesn't use any UDP.  I've seen precedent for `HandleIp6Receive` but that doesn't seem right for all cases either, as we don't use IPv6 in our TCAT over BLE case.

What about `HandleMsgReceive` ?

The latest code line 255-256 has this call: 

```
// Cannot call Receive(..) directly because Setup(..) and mState are private
    mTls.HandleUdpReceive(*message, messageInfo);
```